### ISSUE TITLE
Using "when Foo, Bar" in case when expression behaves unexpectedly

### ISSUE THREAD

**Issue by Sija** [crystal-lang/crystal#3091]
I've tried to illustrate my problem in below snippet:
```crystal
def foo(value)
  case value
  when Int, Float, Time::Span
    pp value.seconds
  end
end
# Raises compile-time exception
foo "bar"
```
https://carc.in/#/r/15gp

**Comment by asterite** [comment]
This is a kind of duplicate of #2464

The above case is rewritten to:

```cr
if value.is_a?(Int) || value.is_a?(Float) || value.is_a?(Time::Span)
  pp value.seconds
end
```

But because `||` doesn't have any flow logic, the compiler can't deduce anything for the type of `value` (it's not restricted)

**Comment by jhass** [comment]
Given this will be solved by solving #2464 then, I'd say we can close this.

**Comment by asterite** [comment]
Sure. I also will try to tackle this, I think it should be easy to implement.
